--- kernel/snd-go7007.c.orig	2007-03-21 13:51:54.000000000 -0400
+++ kernel/snd-go7007.c	2007-09-06 23:45:25.000000000 -0400
@@ -22,6 +22,10 @@
 #include <linux/init.h>
 #if LINUX_VERSION_CODE <= KERNEL_VERSION(2,6,17)
 #include <linux/config.h>
+#else
+#if LINUX_VERSION_CODE >= KERNEL_VERSION(2,6,21)
+#include <linux/spinlock.h>
+#endif
 #endif
 #include <linux/delay.h>
 #include <linux/sched.h>
@@ -57,6 +61,18 @@
 MODULE_PARM_DESC(index, "ID string for the go7007 audio driver");
 MODULE_PARM_DESC(index, "Enable for the go7007 audio driver");
 
+#if LINUX_VERSION_CODE >= KERNEL_VERSION(2,6,21)
+typedef struct snd_card snd_card_t;
+typedef struct snd_pcm snd_pcm_t;
+typedef struct snd_pcm_substream snd_pcm_substream_t;
+typedef struct snd_pcm_runtime snd_pcm_runtime_t;
+typedef struct snd_pcm_hardware snd_pcm_hardware_t;
+typedef struct snd_pcm_hw_params snd_pcm_hw_params_t;
+typedef struct snd_pcm_ops snd_pcm_ops_t;
+typedef struct snd_device snd_device_t;
+typedef struct snd_device_ops snd_device_ops_t;
+#endif
+
 struct go7007_snd {
 	snd_card_t *card;
 	snd_pcm_t *pcm;
